### PR TITLE
Bug fix: invalid ouput_variable entry in bmi_multi module

### PR DESCRIPTION
**Pull request by stcui007**
Fix invalid entry error that causes incorrect output header

## Additions
 -

## Removals

-

## Changes

src/realizations/catchment/Bmi_Multi_Formulation.cpp
test/realizations/catchments/Bmi_Multi_Formulation_Test.cpp

## Testing

1.

## Notes

-

## Todos

-

## Checklist

- [ x] PR has an informative and human-readable title
- [x ] Changes are limited to a single goal (no scope creep)
- [x ] Code can be automatically merged (no conflicts)
- [x ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ x] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

## Testing checklist (automated report can be put here)

1.

### Target Environment support

- [x ] Linux

### CODE REVIEW

File: src/realizations/catchment/Bmi_Multi_Formulation.cpp [hunk 1]
```diff
@@ -303,6 +303,23 @@ string Bmi_Multi_Formulation::get_output_line_for_timestep(int timestep, std::st
...
 
+        std::cout << "Multi: timestep = " << timestep << std::endl;
+        for (int i = 0; i < output_var_names.size(); ++i) {
+            std::cout << "Multi: output_var_names," << output_var_names[i] << std::endl;
+        }
+
+        // Remove invalid output variable name entry in the globle bmi-multi module
+        std::vector<std::string>::iterator it;
+        int last_index = modules.size() - 1;
+        auto module = modules[last_index];
```
Comment (mattw-nws) [inline src/realizations/catchment/Bmi_Multi_Formulation.cpp:314]: Actually this is wrong. The list of valid variable names for a multi_bmi formulation is _not_ found just in the last BMI module in the list; rather it is the union of all names of all BMI modules, including the forcing DataProvider. (Using the outputs from the last module is only relevant as the _default_ set of names used when nothing is explicitly specified, as done further down in this function.)

Perhaps this is why the test below was disabled? Does it attempt to retrieve a variable that is not from the last BMI module in the multi_bmi?

**Review by mattw-nws — CHANGES_REQUESTED** [review]
Beyond what's discussed in the inline review comments, the approach here is actually done in the wrong place, for several reasons.

First, while #535 specifically addresses a mult_bmi formulation, the `output_variables` configuration can be applied to any kind of formulation and the problematic result behavior (headers mismatching the data columns) will occur for other formulation types also--so this needs to be tackled in Bmi_Module_Formulation, somehow, even if it requires common code/a method in that class that is then called somewhere in BMI_[language]_Formulation subclasses. The bmi_multi case might require extra consideration/code because of the unique way that a list of valid variable names is derived (not from a single BMI module), but still the issue affects non-multi formulations.

Also, as discussed in the code comments, this should not be checked every time an output line is processed. First because that will run far more times than necessary, and second because it will only get checked after the simulation has started running at least one timestep. I believe all of the formulation types pre-compute the list of available variables when the formulation is created--this would be the much better place to check if the custom `output_variables` list is valid.

**Comment by stcui007** [comment]
First, I agree that the code as it is currently written runs every time
step. That can be easily fixed.

Also, in its current form in the BMI_Multi_formulation.cpp, it catches the
last module's output header in config jason file for output. I think that's
the module the get_output_line_timestep() uses for writing output. I have
tested and it seems to be the case.

The reason that test 3_a (which is for python) is currently disabled is
because the Bmi_Multi_Formulation_Test.cpp has a different output variable
list than that in the extern/test_bmi_py. That triggers an exception to be
thrown by design in Bmi_Multi_Formulation.cpp.

**Comment by mattw-nws** [comment]
> Also, in its current form in the BMI_Multi_formulation.cpp, it catches the last module's output header in config jason file for output. I think that's the module the get_output_line_timestep() uses for writing output.

I mentioned this in the last review comment, but it captures the last module's output fields as the _default_ set of variables if nothing else is explicitly specified (via `output_variables` on the multi_bmi itself). But this needs to validate if all the variables specified in `output_variables` are in the set of _valid_ variables for the formulation, not if they are all in the _default_ set of variables.

As a concrete example, `output_variables` could contain `water_potential_evaporation_flux`, which comes from the PET module, which is not the _last_ BMI module in the example realization config in #535, but would be a valid option to have in `output_variables`--the code in this PR would throw an exception in this case. Also, it would be valid to include `atmosphere_water__liquid_equivalent_precipitation_rate` in `output_variables`, which typically comes from the forcing data, not from the last BMI module--this code would throw an exception in that case also.

File: src/realizations/catchment/Bmi_Multi_Formulation.cpp (hunk 1 above)
Comment (stcui007) [reply src/realizations/catchment/Bmi_Multi_Formulation.cpp:314]: It is possible we may get an incorrect header in the output if we check all the modules.

Comment (mattw-nws) [reply src/realizations/catchment/Bmi_Multi_Formulation.cpp:314]: How? _Checking_ all of the modules to see if a given variable is valid does not cause all of the modules' outputs to appear in the output--what causes them to appear in the output is if the user has explicitly called for them using `output_variables`.

**Pull request closed without merging**